feat: Upgrade react-native-vision-camera to v4 with stable Skia Frame Processors (& supporting libraries)

ttuck303 opened the pull request:
Found your amazing tutorial (https://mrousavy.com/blog/VisionCamera-Pose-Detection-TFLite) and saw that it was a little out of date. Now that we have Skia frame processors back in Vision Camera 4, I figured we could update the source code. 

Updates libraries:
"@shopify/react-native-skia": "^0.1.202" -> "^1.2.3"
"react-native-fast-tflite": "^0.2.0" -> "^1.2.0"
"react-native-vision-camera": "^3.0.0-rc.7" -> "^4.0.1"
"react-native-worklets-core": "^0.2.0" -> "^2.1.1"

Adds:
"react-native-reanimated": "^3.8.1"
"vision-camera-resize-plugin": "^2.1.1" (more reliable for resizing)

Tested on iPhone 12 Pro:
https://github.com/mrousavy/VisionCameraSkiaDemo/assets/1236380/55420512-bebb-4d5a-8477-f52bd9c7b32d


Two possible concerns:
1. the pose scaling seems overly large / not connected to the head
2. (possibly related to 1) hard coded the camera width to 1080, since original format code was not working for me at the time

Thanks for all you do for the RN Community!

mrousavy commented:
amazing thank you!

mrousavy commented on src/App.tsx:
```diff
@@ -22,14 +21,9 @@ const VIEW_WIDTH = Dimensions.get('screen').width;
-    [device],
-  );
-  console.log(format?.videoWidth, format?.videoHeight);
+  const devices = useCameraDevices();
+  const device = devices.find(d => d.position === 'front');
```
Please simply use `useCameraDevice('front')` here - it's more efficient.

mrousavy commented on package.json:
```diff
@@ -10,17 +10,20 @@
+    "react-native-reanimated": "^3.8.1",
+    "react-native-vision-camera": "^4.0.1",
+    "react-native-worklets-core": "^1.1.1",
+    "vision-camera-resize-plugin": "^2.1.1"
```
Please upgrade to latest 3.x.x, this fixes Android

mrousavy reviewed (changes requested):
amazing thank you - just two minor changes

ttuck303 commented:
- updated libraries and hooks as requested
- I believe the scaling issue in the Y-axis is due to cropping the rectangular image to a square for the pose estimation model, then scaling the coordinates from the model to the rectangular shape without accounting for the cropped area. But not a big deal for this demo app
- ios working
- android - updated minSDK to 26 and targetSDK to 34 for frame processors (did not push yet), compiled but got runtime error `"Frame has invalid PixelFormat! Only YUV_420_888 is supported. Did you set pixelFormat=\"yuv\"?"` - I will look at that next 😎

mrousavy commented:
Thanks! Yea you just need to pass `pixelFormat="yuv"` to the Camera.

mrousavy commented on src/App.tsx:
```diff
@@ -64,7 +58,7 @@ function App(): JSX.Element {
   }
 
   // to get from px -> dp since we draw in the camera coordinate system
-  const SCALE = (format?.videoWidth ?? VIEW_WIDTH) / VIEW_WIDTH;
```
Please don't hardcode this. This is not a good idea.

Above you removed the code that selected a Format, let's keep using a Format so we know video size in advance.

ttuck303 commented:
Updated with requested changes:
- added back dynamic scale
- added pixel format based on device platform
- updated android minsdk and compile sdk to required versions
- added rotation hack to properly orient Android


Something is still incorrect with Android:
- if I leave orientation at 0 in resize plugin, the final image is rotated 90
- if I rotate the orientation 270 in the resize plugin, the final image is properly oriented, but squished. 
Hoping its an obvious fix. 

Running on iOS:
https://github.com/mrousavy/VisionCameraSkiaDemo/assets/1236380/084d517e-5606-4c6e-8498-8c8eef07440c

Running on Android with 270deg rotation:
https://github.com/mrousavy/VisionCameraSkiaDemo/assets/1236380/92117f84-a0a9-4bd4-aa16-a90232c28a57

Running on Android with 0deg rotation:
https://github.com/mrousavy/VisionCameraSkiaDemo/assets/1236380/268c0fbc-28b6-48af-9b78-02dfc248036c

mrousavy commented:
Awesome work man! Thank you
